pandas/core/util/hashing.py

```diff
@@ -299,3 +310,63 @@ def hash_array(vals, encoding='utf8', hash_key=None, categorize=True):
     vals *= np.uint64(0x94d049bb133111eb)
     vals ^= vals >> 31
     return vals
+
+
+def _hash_scalar(val, encoding='utf8', hash_key=None):
+    """
```

Comment:
Yes, I know, but this was mainly to test. 
And from that it, this makes it a lot faster than using the `hash_array` (the commented part in `hash_tuple`). So not sure how to solve that. In principle I can put the common parts in helper functions (eg the redistributing part), but for most of it it is not possible as there are slight differences.